### ISSUE TITLE
Export to DFC prototype fails on iterating catalog items

### ISSUE THREAD

**Issue by mkllnk** [datafoodconsortium/connector-ruby#11]
The prototype shows the error:

> Error: platformUserDereferences.dfc-b:affiliates.dfc-b:manages is not iterable

The generated JSON:
```json
{
  "@context": "http://static.datafoodconsortium.org/ontologies/context.json",
  "@graph": [
    {
      "@id": "https://staging.coopcircuits.fr/api/dfc-v1.6/persons/21208",
      "@type": "dfc-b:Person",
      "http://static.datafoodconsortium.org/ontologies/DFC_BusinessOntology.owl#affiliates": "https://staging.coopcircuits.fr/api/dfc-v1.6/enterprises/2731"
    },
    {
      "@id": "https://staging.coopcircuits.fr/api/dfc-v1.6/enterprises/2731",
      "@type": "dfc-b:Enterprise",
      "dfc-b:hasName": "",
      "dfc-b:hasDescription": "",
      "dfc-b:VATnumber": "",
      "http://static.datafoodconsortium.org/ontologies/DFC_BusinessOntology.owl#supplies": [
        "https://staging.coopcircuits.fr/api/dfc-v1.6/enterprises/2731/supplied_products/56790",
        "https://staging.coopcircuits.fr/api/dfc-v1.6/enterprises/2731/supplied_products/56789"
      ],
      "http://static.datafoodconsortium.org/ontologies/DFC_BusinessOntology.owl#manages": [
        "https://staging.coopcircuits.fr/api/dfc-v1.6/enterprises/2731/catalog_items/56790",
        "https://staging.coopcircuits.fr/api/dfc-v1.6/enterprises/2731/catalog_items/56789"
      ]
    },
    {
      "@id": "https://staging.coopcircuits.fr/api/dfc-v1.6/enterprises/2731/catalog_items/56790",
      "@type": "dfc-b:CatalogItem",
      "http://static.datafoodconsortium.org/ontologies/DFC_BusinessOntology.owl#references": "https://staging.coopcircuits.fr/api/dfc-v1.6/enterprises/2731/supplied_products/56790",
      "dfc-b:sku": "VAR",
      "dfc-b:stockLimitation": 0,
      "http://static.datafoodconsortium.org/ontologies/DFC_BusinessOntology.owl#offeredThrough": "https://staging.coopcircuits.fr/api/dfc-v1.6/enterprises/2731/offers/56790"
    },
    {
      "@id": "https://staging.coopcircuits.fr/api/dfc-v1.6/enterprises/2731/catalog_items/56789",
      "@type": "dfc-b:CatalogItem",
      "http://static.datafoodconsortium.org/ontologies/DFC_BusinessOntology.owl#references": "https://staging.coopcircuits.fr/api/dfc-v1.6/enterprises/2731/supplied_products/56789",
      "dfc-b:sku": "BAN",
      "dfc-b:stockLimitation": 0,
      "http://static.datafoodconsortium.org/ontologies/DFC_BusinessOntology.owl#offeredThrough": "https://staging.coopcircuits.fr/api/dfc-v1.6/enterprises/2731/offers/56789"
    },
    {
      "@id": "https://staging.coopcircuits.fr/api/dfc-v1.6/enterprises/2731/supplied_products/56790",
      "@type": "dfc-b:SuppliedProduct",
      "dfc-b:name": "Banana updated",
      "dfc-b:description": "\r\n\r\n",
      "dfc-b:hasQuantity": {
        "@type": "dfc-b:QuantitativeValue",
        "http://static.datafoodconsortium.org/ontologies/DFC_BusinessOntology.owl#hasUnit": "dfc-m:Piece",
        "dfc-b:value": 3
      },
      "dfc-b:alcoholPercentage": 0,
      "dfc-b:lifetime": "",
      "dfc-b:usageOrStorageCondition": "",
      "dfc-b:totalTheoreticalStock": 0
    },
    {
      "@id": "https://staging.coopcircuits.fr/api/dfc-v1.6/enterprises/2731/supplied_products/56789",
      "@type": "dfc-b:SuppliedProduct",
      "dfc-b:name": "Banana updated",
      "dfc-b:description": "\r\n\r\n",
      "dfc-b:hasQuantity": {
        "@type": "dfc-b:QuantitativeValue",
        "http://static.datafoodconsortium.org/ontologies/DFC_BusinessOntology.owl#hasUnit": "dfc-m:Piece",
        "dfc-b:value": 0
      },
      "dfc-b:alcoholPercentage": 0,
      "dfc-b:lifetime": "",
      "dfc-b:usageOrStorageCondition": "",
      "dfc-b:totalTheoreticalStock": 0
    },
    {
      "@id": "https://staging.coopcircuits.fr/api/dfc-v1.6/enterprises/2731/offers/56790",
      "@type": "dfc-b:Offer",
      "dfc-b:price": 0,
      "dfc-b:stockLimitation": 3
    },
    {
      "@id": "https://staging.coopcircuits.fr/api/dfc-v1.6/enterprises/2731/offers/56789",
      "@type": "dfc-b:Offer",
      "dfc-b:price": 0,
      "dfc-b:stockLimitation": 0
    }
  ]
}
```

**Comment by RaggedStaff** [comment]
Looks like the proto is missing the links between Person & Enterprise (dfc-b.affiliates) and CatalogItem & SuppliedProduct (dfc-b.references). For some reason these are using the full path to the ontology, not the prefix... that's probably the issue... although the json is valid, could be a bug for the proto... what do you think @lecoqlibre ?

**Comment by lecoqlibre** [comment]
Maybe it's because of the `affiliates` property is not an array. Maybe you could try to change:
```
"http://static.datafoodconsortium.org/ontologies/DFC_BusinessOntology.owl#affiliates": "https://staging.coopcircuits.fr/api/dfc-v1.6/enterprises/2731"
```
to:
```
"http://static.datafoodconsortium.org/ontologies/DFC_BusinessOntology.owl#affiliates": [
    "https://staging.coopcircuits.fr/api/dfc-v1.6/enterprises/2731"
]
```
using an array @mkllnk.

**Comment by mkllnk** [comment]
> Maybe it's because of the affiliates property is not an array.

But I'm assigning an array:

```rb
person.affiliatedOrganizations.class
=> Array
```

I'm mapping the OFN enterprises to DFC enterprises and keep it as array:
```rb
class PersonBuilder < DfcBuilder
  def self.person(user)
    DataFoodConsortium::Connector::Person.new(
      urls.person_url(user.id),
      firstName: user.bill_address&.firstname,
      lastName: user.bill_address&.lastname,
      affiliatedOrganizations: enterprises(user),
    )
  end

  def self.enterprises(user)
    user.enterprises.map do |enterprise|
      EnterpriseBuilder.enterprise(enterprise)
    end
  end
end
```
https://github.com/openfoodfoundation/openfoodnetwork/blob/50463615dae8999b9843308ac558809182d861f3/engines/dfc_provider/app/services/person_builder.rb#L3-L18

I just tried your example from the Readme and it doesn't produce an array either:
```json
{
  "@context": "http://static.datafoodconsortium.org/ontologies/context.json",
  "@graph": [
    {
      "@id": "https://myplatform.com/p",
      "@type": "dfc-b:Person",
      "dfc-b:hasAddress": "https://myplatform.com/a",
      "dfc-b:firstName": "firstName",
      "dfc-b:familyName": "lastName",
      "http://static.datafoodconsortium.org/ontologies/DFC_BusinessOntology.owl#affiliates": "https://myplatform.com/e"
    },
    {
      "@id": "https://myplatform.com/a",
      "@type": "dfc-b:Address",
      "dfc-b:hasStreet": "street",
      "dfc-b:hasPostalCode": "postalCode",
      "dfc-b:hasCity": "city",
      "dfc-b:hasCountry": "country"
    },
    {
      "@id": "https://myplatform.com/e",
      "@type": "dfc-b:Enterprise",
      "dfc-b:hasName": "name",
      "dfc-b:hasDescription": "description",
      "dfc-b:VATnumber": "vatNumber",
      "http://static.datafoodconsortium.org/ontologies/DFC_BusinessOntology.owl#defines": "https://myplatform.com/cc",
      "http://static.datafoodconsortium.org/ontologies/DFC_BusinessOntology.owl#supplies": "https://myplatform.com/sp",
      "http://static.datafoodconsortium.org/ontologies/DFC_BusinessOntology.owl#manages": "https://myplatform.com/ci"
    },
    {
      "@id": "https://myplatform.com/cc",
      "@type": "dfc-b:CustomerCategory",
      "dfc-b:description": "description"
    },
    {
      "@id": "https://myplatform.com/sp",
      "@type": "dfc-b:SuppliedProduct",
      "dfc-b:name": "name",
      "dfc-b:description": "description",
      "http://static.datafoodconsortium.org/ontologies/DFC_BusinessOntology.owl#hasType": "http://static.datafoodconsortium.org/data/productTypes.rdf#artichoke",
      "dfc-b:hasQuantity": {
        "@type": "dfc-b:QuantitativeValue",
        "http://static.datafoodconsortium.org/ontologies/DFC_BusinessOntology.owl#hasUnit": "dfc-m:Kilogram",
        "dfc-b:value": 1.2
      },
      "dfc-b:alcoholPercentage": 2.6,
      "dfc-b:lifetime": "lifetime",
      "http://static.datafoodconsortium.org/ontologies/DFC_BusinessOntology.owl#hasClaim": [
        "dfc-f:EnergyFree",
        "dfc-f:FatFree",
        "dfc-f:HighFibre"
      ],
      "dfc-b:usageOrStorageCondition": "usageOrStorageConditions",
      "dfc-b:hasNutrientCharacteristic": {
        "@type": "dfc-b:NutrientCharacteristic",
        "http://static.datafoodconsortium.org/ontologies/DFC_BusinessOntology.owl#hasUnit": "dfc-m:Gram",
        "dfc-b:value": 8.47,
        "dfc-b:hasNutrientDimension": "dfc-m:Calcium"
      },
      "dfc-b:hasPhysicalCharacteristic": {
        "@type": "dfc-b:PhysicalCharacteristic",
        "http://static.datafoodconsortium.org/ontologies/DFC_BusinessOntology.owl#hasUnit": "dfc-m:Kilogram",
        "dfc-b:value": 3.25,
        "http://static.datafoodconsortium.org/ontologies/DFC_BusinessOntology.owl#hasPhysicalDimension": "dfc-m:Weight"
      },
      "http://static.datafoodconsortium.org/ontologies/DFC_BusinessOntology.owl#hasGeographicalOrigin": "dfc-f:Normandy",
      "dfc-b:hasCertification": "dfc-f:Demeter",
      "dfc-b:hasNatureOrigin": "dfc-f:PlantOrigin",
      "dfc-b:hasPartOrigin": "dfc-f:PlantOrigin",
      "dfc-b:totalTheoreticalStock": 123
    },
    {
      "@id": "https://myplatform.com/ci",
      "@type": "dfc-b:CatalogItem",
      "http://static.datafoodconsortium.org/ontologies/DFC_BusinessOntology.owl#references": "https://myplatform.com/sp",
      "dfc-b:sku": "sku",
      "dfc-b:stockLimitation": 10,
      "http://static.datafoodconsortium.org/ontologies/DFC_BusinessOntology.owl#offeredThrough": "https://myplatform.com/o"
    },
    {
      "@id": "https://myplatform.com/o",
      "@type": "dfc-b:Offer",
      "dfc-b:price": {
        "@type": "dfc-b:Price",
        "dfc-b:value": 12.78,
        "dfc-b:VATrate": 5.22,
        "http://static.datafoodconsortium.org/ontologies/DFC_BusinessOntology.owl#hasUnit": "dfc-m:Euro"
      },
      "dfc-b:stockLimitation": 52,
      "dfc-b:offeredTo": "https://myplatform.com/cc"
    }
  ]
}
```

**Comment by lecoqlibre** [comment]
Yes it's because with JSON-LD compacted form, an array containing a single element is optimized to a string (see the [specification](https://www.w3.org/TR/json-ld11/#lists-and-sets)).

I think the prototype should handle this properly. I will ask @simonLouvet.

It might be connected to issue #7.

**Comment by mkllnk** [comment]
> Which prototype are you using?

proto.datafoodconsortium.org

> Can you try with prefixing in your JSON the unprefixed properties like manages with dfc-b?

How do I do that? I'm just using the DFC Connector methods.

```rb
enterprise.catalogItems = catalog_items
```

**Comment by lecoqlibre** [comment]
I just released the [alpha.6](https://rubygems.org/gems/datafoodconsortium-connector/versions/1.0.0.pre.alpha.6) version of the connector. Updating your setup should fix the problem @mkllnk.

The problem was due to the context. When we generate the JSON in the connector the properties were not prefixed yet: we had `http://static.datafoodconsortium.org/ontologies/DFC_BusinessOntology.owl#supplies` instead of `dfc-b:supplies`.

To make it work with the public DFC's context we should define a term with an `@id` like:
```
"supplies": {
      "@id": "http://static.datafoodconsortium.org/ontologies/DFC_BusinessOntology.owl#supplies",
      "@type":"@id"
}
```

As we won't change the public DFC's context, we are now prefixing the properties before to export them with the public DFC's context.
